Spawning tweaks.

MistakeNot4892 opened the pull request:
## Description of changes
- Various global latejoin lists are replaced with spawnpoint decls.
- Observer spawnpoints are now a spawnpoint decl.
- Name-based backup decl finding if uid fails has been generalized in a proc on decls_repository.
- Spawnpoint selection in prefs is now non-modal.

## Why and what will this PR improve
Nicer code, easier to override/modify downstream or in a modpack.

## Authorship
Myself.

## Changelog
Nothing player-facing.

Andrew-Fall commented on code/controllers/subsystems/jobs.dm:
```diff
@@ -505,7 +505,7 @@ SUBSYSTEM_DEF(jobs)
 		else
 			var/decl/spawnpoint/spawnpoint = job.get_spawnpoint(H.client)
-			H.forceMove(pick(spawnpoint.turfs))
+			H.forceMove(length(spawnpoint.spawn_turfs) ? pick(spawnpoint.spawn_turfs) : locate(1,1,1))
```
DEFAULTPICK could be used instead

Andrew-Fall commented on code/modules/mob/observer/ghost/ghost.dm:
```diff
@@ -61,11 +61,15 @@ var/global/list/image/ghost_sightless_images = list() //this is a list of images
+			var/decl/spawnpoint/spawn_data = all_spawns[spawn_type]
+			if(spawn_data.ghost_can_spawn && length(spawn_data.spawn_turfs))
+				spawn_locs |= spawn_data.spawn_turfs
+		T = length(spawn_locs) ? pick(spawn_locs) : get_respawn_loc()
```
DEFAULTPICK

Andrew-Fall replied:
Also I feel as though it would be better to use get_respawn_loc() and have that proc do checks on either flags or using another proc that takes the mob as a parameter.

Andrew-Fall commented on code/modules/mob/new_player/new_player.dm:
```diff
@@ -360,7 +363,7 @@ INITIALIZE_IMMEDIATE(/mob/new_player)
 			job = SSjobs.get_by_title(global.using_map.default_job_title)
 		var/decl/spawnpoint/spawnpoint = job.get_spawnpoint(client, client.prefs.ranks[job.title])
-		spawn_turf = pick(spawnpoint.turfs)
+		spawn_turf = length(spawnpoint.spawn_turfs) ? pick(spawnpoint.spawn_turfs) : get_respawn_loc()
```
DEFAULTPICK

Andrew-Fall commented on code/datums/repositories/decls.dm:
```diff
@@ -40,6 +40,19 @@ var/global/repository/decls/decls_repository = new
 	RETURN_TYPE(/decl)
 	. = get_decl(fetched_decl_ids[decl_id], validate_decl_type)
 
+/repository/decls/proc/get_decl_by_id_or_var(var/decl_id, var/decl_prototype, var/check_var = "name")
```
Given the code this is replacing, I feel like this should have a comment saying it's for grandfathered in items and shouldn't be used for new code without good reason.

MistakeNot4892 replied:
Comments added.

out-of-phaze commented on code/controllers/subsystems/jobs.dm:
```diff
@@ -505,7 +505,7 @@ SUBSYSTEM_DEF(jobs)
 		else
 			var/decl/spawnpoint/spawnpoint = job.get_spawnpoint(H.client)
-			H.forceMove(pick(spawnpoint.turfs))
+			H.forceMove(length(spawnpoint.spawn_turfs) ? pick(spawnpoint.spawn_turfs) : locate(1,1,1))
```
```suggestion
			H.forceMove(DEFAULTPICK(spawnpoint.spawn_turfs, locate(1,1,1)))
```
I kind of feel like we should log if it doesn't have turfs, though. Is that done elsewhere?

MistakeNot4892 replied:
There's a unit test to validate that all spawnpoints that are either in the allowed list or forced by jobs have turfs to spawn on.

out-of-phaze reviewed (changes requested):
Seconding the requests for changes from Andrew. I'm also not really sure what the purpose of `get_respawn_loc()` is. Is it just... *any* turf you could spawn on? The name doesn't make it clear and there's no comments, either.

MistakeNot4892 commented:
Okay, this has spiraled a bit.
- Replaced `spawn_turfs` being used directly with a proc so I can override it on my fork and check prefs/culture stuff.
- Added a flag system instead of the bools and made the random spawn point proc take a flag to filter results.
- Moved the gateway onto Exodus since that's the only map that uses it.
- Made the default allowed spawns just Arrivals.

out-of-phaze commented on code/unit_tests/map_tests.dm:
```diff
@@ -299,19 +299,70 @@
+	var/list/check_spawn_flags = list(
+		"SPAWN_FLAG_PRISONERS_CAN_SPAWN" = SPAWN_FLAG_PRISONERS_CAN_SPAWN,
+		"SPAWN_FLAG_JOBS_CAN_JOIN"       = SPAWN_FLAG_JOBS_CAN_SPAWN,
+		"SPAWN_FLAG_PERSISTENCE_CAN_USE" = SPAWN_FLAG_PERSISTENCE_CAN_SPAWN
```
```suggestion
		"SPAWN_FLAG_JOBS_CAN_SPAWN"       = SPAWN_FLAG_JOBS_CAN_SPAWN,
		"SPAWN_FLAG_PERSISTENCE_CAN_SPAWN" = SPAWN_FLAG_PERSISTENCE_CAN_SPAWN
```
I'm not sure why these don't match?

MistakeNot4892 replied:
Oh, I renamed the flags at some point. The string is just for display purposes but I'll amend it.

out-of-phaze commented on code/modules/client/preferences_spawnpoints.dm:
```diff
@@ -1,14 +1,50 @@
+	/// Determines validity for get_random_spawn_turf()
+	var/spawn_flags = (SPAWN_FLAG_GHOSTS_CAN_SPAWN | SPAWN_FLAG_JOBS_CAN_SPAWN | SPAWN_FLAG_PRISONERS_CAN_SPAWN | SPAWN_FLAG_PERSISTENCE_CAN_SPAWN)
+	/// List of turfs to spawn on. Retrieved via get_spawn_turfs().
+	var/list/_spawn_turfs
```
I think this should be `VAR_PRIVATE/list/_spawn_turfs` or `VAR_PROTECTED/list/_spawn_turfs`. Unless you have a use-case downstream/in a modpack for things other than `/decl/spawnpoint` and its subtypes modifying `_spawn_turfs` directly rather than through helpers.
